state/application.go

```diff
@@ -2400,11 +2400,12 @@ func (a *Application) addUnitOps(
 		// has the special juju- prefix to its name, then bypass the machineID
 		// empty check.
 		if args.machineID != "" && a.st.IsController() {
-			curl, err := charm.ParseURL(*a.doc.CharmURL)
+			ch, _, err := a.Charm()
 			if err != nil {
 				return "", nil, errors.Trace(err)
 			}
-			if !strings.HasPrefix(curl.Name, "juju-") {
+			name := ch.Meta().Name
```

Comment:
Meta().Name is not a required value. Please add a comment, that this works because the juju-controller charm has the Name set.

Reply:
If it's not set we could probably come up with a temp one.